doc/design/dynamic_rnn_beamsearch.md

```diff
@@ -0,0 +1,120 @@
+
+One of the mature DL platforms, TensorFlow, offers a mass of operators for mathematical and conditional operations,
+in that way, TF users can simulate the Python-way and uses some dynamic RNN, encoder or decoder to build an RNN model represented as a graph and the model supports large-scale computation.
+
+That is a good way to follow, by turning a complex algorithm into the combination of several highly-reusable operators, make our framework more flexible.
+
+In this documentation, we will first give a glimpse of the dynamic encoder and dynamic decoder in TensorFlow, then propose a general scheme of our dynamic RNN's design.
+
+## Dynamic RNN in TensorFlow
+
+### static RNN
+The most basic of an RNN is `RNNCell`, it has a `__call__` method that supports `output, next_state = call(input, state).`
+the basic way to define a static RNN like follows
+
+```python
+dataset = tf.placeholder(...)
+lstm = tf.contrib.rnn.BasicLSTMCell(sate_size)
+
+for step_input in dataset:
+    output, state = lstm(step_input, state)
+```
+
+the code above will build an RNN with `step_input` and last `state` as inputs, the operators in `BasicLSTMCell` will be copied each step, 
+and a very wide and big static graph will be generated.
+
+<p align="center">
+<img src="./images/dynamic_rnn1.png"/>
+</p>
+
+### dynamic RNN
+
+With the `while_loop` and some other conditional operators, one can turn the static RNN into a dynamic RNN. 
+Some code snippets as follows
+
+```python
+_, output_final_ta, final_state = while_loop(
+    cond=lambda time, *_: time < time_steps,
+    body=_time_step,
+    loop_vars=(time, output_ta, state),
+    parallel_iteratioons=parallel_iterations)
+```
+
+the `cond` is a conditional method that determines whether to continue the loop,
+the `body` 
+
+For more details, reference https://www.tensorflow.org/api_docs/python/tf/nn/dynamic_rnn
+
+
+
+## Design of Dynamic RNN
+
+Due to the different underlying architecture, it is hard to copy the design of dynamic operators from TensorFlow, but the ideas behind design are worth learning, that is
+
+1. introducing some general conditional operators to simulate a programming language,
+2. a `TensorArray` to store inputs and outputs in RNN's time steps.
+
+```python
+input = pd.data(xxx) # a data layer
+# if dynamic, vararray will extend dynamicly if more entries needed
+states = pd.VarArray(dynamic=True)
+step_inputs = pd.VarArray().segment(input) # segment is an operator
+
+# a scalar
+time = pd.Variable(shape=[0], dtype=pd.int32)
+# segmented steps, size() is an operator which takes an VarArray as input and
+# output a scalar tensor.
+num_steps = step_inputs.size()
+
+W = pd.parameter()
+U = pd.parameter()
+
+boot_state = some_op()
+
+whl = pd.while_loop()
+
+# by default, exec_immediately is True, that means when exit the with-statement of
+# whl.loop(), whl() will be called
+with whl.loop(exec_immediately=True):
+    time_ = whl.loop_var(time)
+    to_stop = pd.less_than(time_, num_steps)
```

Comment:
iterating over the input sequence can be a default condition check for dynamic RNN training/testing.

For beam training/generation more complicated condition is required.